examples/diagram.dx

```diff
@@ -0,0 +1,193 @@
+
+def scale (s:Float) : (Diagram -> Diagram) =
+  applyTransformation ( \(x,y). (s * x, s * y) ) \geom. case geom of
+    Spot          -> Spot
+    Circle r      -> Circle (s * r)
+    Rectangle w h -> Rectangle (s * w) (s * h)
+    Line (x, y)   -> Line (s * x, s * y)
+
+def moveXY ((offX, offY) : Point) : (Diagram -> Diagram) =
+  applyTransformation (\(x,y). (x + offX, y + offY) ) id
+
+def singletonDefault (geom:Geom) : Diagram =
+  MkDiagram $ AsList _ [(defaultAttributes, (0.0, 0.0), geom)]
+
+-- TODO: should really allow nullary functions with `def`
+spot                           : Diagram = singletonDefault Spot
+def circle (r:Float)           : Diagram = singletonDefault $ Circle r
+def rect   (w:Float) (h:Float) : Diagram = singletonDefault $ Rectangle w h
+def line   (p:Point)           : Diagram = singletonDefault $ Line p
+
+def updateAttributes (d:Diagram) (update: Attributes -> Attributes) : Diagram =
+  (MkDiagram (AsList _ objs)) = d
+  MkDiagram $ AsList _ for i.
+    (attr       , geoms) = objs.i
+    (update attr, geoms)
+
+def setFill   (c:Color) (d:Diagram) : Diagram = updateAttributes d $ setAt #fill   (Just c)
+def setStroke (c:Color) (d:Diagram) : Diagram = updateAttributes d $ setAt #stroke (Just c)
+def removeFill          (d:Diagram) : Diagram = updateAttributes d $ setAt #fill   Nothing
+def removeStroke        (d:Diagram) : Diagram = updateAttributes d $ setAt #stroke Nothing
+def setLW     (w:Int)   (d:Diagram) : Diagram = updateAttributes d $ setAt #lw     w
+
+-- === serialization ===
+
+def quote (s:String) : String = str "\"" <> s <> str "\""
+
+def (<+>) (s1:String) (s2:String) : String = s1 <> str " " <> s2
+
+def angleBrackets (s:String) : String = str "<" <> s <> str "/>"
+
+def tagBrackets (tag:String) (s:String) : String =
+  str "<" <> tag <> str ">" <> s <> str "</" <> tag <> str ">"
+
+def tagBracketsAttr (tag:String) (attr:String) (s:String) : String =
+  str "<" <> tag <+> attr <> str ">" <> s <> str "</" <> tag <> str ">"
+
+def makeAttr (attr:String) (val:String) : String =
+  attr <> str "=" <> quote val
+
+def colorStr (c: Maybe Color) : String =
+  case c of
+    Nothing -> str "none"
+    Just s  -> s
+
+def attrString (attr:Attributes) : String =
+  (  makeAttr (str "stroke")       (colorStr $ getAt #stroke attr)
+  <> makeAttr (str "fill")         (colorStr $ getAt #fill   attr)
+  <> makeAttr (str "stroke-width") (show $ getAt #lw attr))
+
+def spotAttrString (attr:Attributes) : String =
+  (  makeAttr (str "stroke")       (colorStr $ getAt #stroke attr)
+  <> makeAttr (str "fill")         (colorStr $ getAt #stroke attr)
+  <> makeAttr (str "stroke-width") (show $ getAt #lw attr))
+
+def renderGeom (attr:Attributes) ((x,y):Point) (geom:Geom) : String =
+  case geom of
+    Spot ->
+     tagBracketsAttr (str "g") (spotAttrString attr) $ angleBrackets $
+       (str "circle" <+>
+        str "cx=" <> quote (show x) <>
+        str "cy=" <> quote (show y) <>
+        str "r=\"1\"")
```

Comment:
Seems a bit weird to have a `tagBracketsAttr` helper for some tags but directly build the string for others. Maybe we could be consistent, and either build it manually everywhere or have a helper for this case too?